sqlmesh/core/macros.py

```diff
@@ -388,6 +423,75 @@ def var(self, var_name: str, default: t.Optional[t.Any] = None) -> t.Optional[t.
         """Returns the value of the specified variable, or the default value if it doesn't exist."""
         return (self.locals.get(c.SQLMESH_VARS) or {}).get(var_name.lower(), default)
 
+    def _coerce(self, expr: exp.Expression, typ: t.Any, strict: bool = False) -> t.Any:
+        """Coerces the given expression to the specified type on a best-effort basis."""
+        base_err_msg = f"Failed to coerce expression '{expr}' to type '{typ}'."
+        try:
+            if typ is None or typ is t.Any:
+                return expr
+            base = t.get_origin(typ) or typ
+            # We need to handle t.Union first since we cannot use isinstance with it
+            if base in (t.Union, types.UnionType):
+                for branch in t.get_args(typ):
+                    try:
+                        return self._coerce(expr, branch, True)
+                    except Exception:
+                        pass
+                raise SQLMeshError(base_err_msg)
+            if isinstance(expr, base):
+                return expr
+            if issubclass(base, exp.Expression):
+                d = Dialect.get_or_raise(self.dialect)
+                into = base if base in d.parser().EXPRESSION_PARSERS else None
+                if into is None:
+                    if isinstance(expr, exp.Literal):
+                        coerced = parse_one(expr.this)
+                    else:
+                        raise SQLMeshError(
+                            f"{base_err_msg} Coercion to {base} requires a literal expression."
+                        )
+                else:
+                    coerced = parse_one(
+                        expr.this if isinstance(expr, exp.Literal) else expr.sql(), into=into
+                    )
+                if isinstance(coerced, base):
+                    return coerced
+                raise SQLMeshError(base_err_msg)
+
+            if base in (int, float, str) and isinstance(expr, exp.Literal):
+                return base(expr.this)
+            if base is bool and isinstance(expr, exp.Boolean):
+                return expr.this
+            if base is str and isinstance(expr, exp.Expression):
+                return expr.sql(self.dialect)
```

Comment:
This seems sketchy. What's the use case for this? IMO I think we should only coerce string literals and identifiers into strings.

Reply:
this seems reasonable to me. for example, you may pass a column or a table into a function, and just care about the string representation

Reply:
Sure, I think this is reasonable for columns, tables and literals. But this logic will apply even if I pass an arbitrary SQL expression, right?

Reply:
without this coercision to string, usage like this is flawed because x will become sqlglot dialect sql